packages/telemetry/src/flight-recorder.js

```diff
@@ -0,0 +1,249 @@
+// @ts-check
+/* global BigUint64Array */
+/// <reference types="ses" />
+
+// https://github.com/Agoric/agoric-sdk/issues/3742#issuecomment-1028451575
+// I'd mmap() a 100MB file, reserve a few bytes for offsets, then use the rest
+// as a circular buffer to hold length-prefixed records. The agd process would
+// keep writing new events into the RAM window and updating the start/end
+// pointers, with some sequencing to make sure the record gets written before
+// the pointer is updated. Then, no mattter how abruptly the process is
+// terminated, as long as the host computer itself is still running, the on-disk
+// file would contain the most recent state, and anybody who reads the file will
+// get the most recent state. The host kernel (linux) is under no obligation to
+// flush it to disk any particular time, but knows when reads happen, so there's
+// no coherency problem, and the speed is unaffected by disk write speeds.
+
+import BufferFromFile from 'bufferfromfile';
+import { promises as fsPromises } from 'fs';
+import path from 'path';
+
+const { details: X } = assert;
+
+export const DEFAULT_CIRCULAR_BUFFER_SIZE = 100 * 1024 * 1024;
+export const DEFAULT_CIRCULAR_BUFFER_FILE = 'flight-recorder.bin';
+export const SLOG_MAGIC = 0x41472d534c4f4721n; // 'AG-SLOG!'
+
+const I_MAGIC = 0 * BigUint64Array.BYTES_PER_ELEMENT;
+const I_ARENA_SIZE = 1 * BigUint64Array.BYTES_PER_ELEMENT;
+const I_CIRC_START = 2 * BigUint64Array.BYTES_PER_ELEMENT;
+const I_CIRC_END = 3 * BigUint64Array.BYTES_PER_ELEMENT;
+const I_ARENA_START = 4 * BigUint64Array.BYTES_PER_ELEMENT;
+
+const initializeCircularBuffer = async (bufferFile, circularBufferSize) => {
```

Comment:
It looks like this function is nominally idempotent (calling it when a suitable file already exists will NOP), but if I do:

```js
await initializeCircularBuffer(filename, 100*MB);
await initializeCircularBuffer(filename, 200*MB);
```

then I think it will grow the file but not update the `ARENA_SIZE` header, so the new space won't be usable.

Not sure if the fix is to assert that `circularBufferSize` matches the existing file size, or to attempt some sort of grow-in-place code (which sounds awful).

Reply:
We assert the arena size below, so all I need to do is bail if there is already a file.